vms/avm/vm.go

```diff
@@ -483,26 +483,85 @@ func (vm *VM) GetUTXOs(
 		limit = maxUTXOsToFetch
 	}
 
-	seen := ids.Set{} // IDs of UTXOs already in the list
+	if paginate {
+		return vm.getPaginatedUTXOs(addrs, startAddr, startUTXOID, limit)
+	}
+	return vm.getAllUTXOs(addrs, startAddr, startUTXOID)
+}
+
+func (vm *VM) getPaginatedUTXOs(addrs ids.ShortSet,
+	startAddr ids.ShortID,
+	startUTXOID ids.ID,
+	limit int,
+) ([]*avax.UTXO, ids.ShortID, ids.ID, error) {
+	seen := ids.New(limit) // IDs of UTXOs already in the list
 	utxos := make([]*avax.UTXO, 0, limit)
 	lastAddr := ids.ShortEmpty
 	lastIndex := ids.Empty
 	addrsList := addrs.List()
 	ids.SortShortIDs(addrsList)
+
+	for _, addr := range addrsList {
+		start := ids.Empty
+		if comp := bytes.Compare(addr.Bytes(), startAddr.Bytes()); comp == -1 { // Skip addresses before [startAddr]
+			continue
+		} else if comp == 0 {
+			start = startUTXOID
+		}
+
+		utxoIDs, err := vm.state.Funds(addr.Bytes(), start, limit) // Get UTXOs associated with [addr]
+		if err != nil {
+			return nil, ids.ShortID{}, ids.ID{}, fmt.Errorf("couldn't get UTXOs for address %s: %w", addr, err)
+		}
+		for _, utxoID := range utxoIDs {
+			if seen.Contains(utxoID) { // Already have this UTXO in the list
+				continue
+			}
+			utxo, err := vm.state.UTXO(utxoID)
+			if err != nil {
+				return nil, ids.ShortID{}, ids.ID{}, fmt.Errorf("couldn't get UTXO %s: %w", utxoID, err)
+			}
+			utxos = append(utxos, utxo)
+			seen.Add(utxoID)
+			lastAddr = addr
+			lastIndex = utxoID
+			limit--
+			if limit <= 0 {
+				break // Found [limit] utxos; stop.
+			}
+		}
+	}
+	return utxos, lastAddr, lastIndex, nil
+}
+
+func (vm *VM) getAllUTXOs(addrs ids.ShortSet,
+	startAddr ids.ShortID,
+	startUTXOID ids.ID,
+) ([]*avax.UTXO, ids.ShortID, ids.ID, error) {
+	seen := ids.New(maxUTXOsToFetch) // IDs of UTXOs already in the list
+	utxos := make([]*avax.UTXO, 0, maxUTXOsToFetch)
+	lastAddr := ids.ShortEmpty
+	lastIndex := ids.Empty
+	addrsList := addrs.List()
+	ids.SortShortIDs(addrsList)
+
 	for _, addr := range addrsList {
 		start := ids.Empty
 		if comp := bytes.Compare(addr.Bytes(), startAddr.Bytes()); comp == -1 { // Skip addresses before [startAddr]
 			continue
 		} else if comp == 0 {
 			start = startUTXOID
 		}
+
 		for {
-			currLimit := limit
-			utxoIDs, err := vm.state.Funds(addr.Bytes(), start, limit) // Get UTXOs associated with [addr]
+			utxoIDs, err := vm.state.Funds(addr.Bytes(), start, maxUTXOsToFetch) // Get UTXOs associated with [addr]
 			if err != nil {
 				return nil, ids.ShortID{}, ids.ID{}, fmt.Errorf("couldn't get UTXOs for address %s: %w", addr, err)
 			}
 			for _, utxoID := range utxoIDs {
+				lastAddr = addr
```

Comment:
Similarly to above, I think this should be moved out to right after the comparing of the `addr`s.

Reply:
Yeah, fundamentally we're implementing two different outcomes on the Paginated and the GetAll.
The Paginated will return `LastAddress : A / LastIndex : UTXO1` while the GetAll will return `LastAddress : B / LastIndex : UTXO1` as the example from here https://github.com/ava-labs/avalanchego/pull/604#discussion_r527634872